src/osd/osd_types.cc

```diff
@@ -1293,14 +1293,116 @@ void pg_pool_t::build_removed_snaps(interval_set<snapid_t>& rs) const
 {
   if (is_pool_snaps_mode()) {
     rs.clear();
-    for (snapid_t s = 1; s <= get_snap_seq(); s = s + 1)
-      if (snaps.count(s) == 0)
-	rs.insert(s);
+    snapid_t last = 0;
+    snapid_t snapseq_max = get_snap_seq();
+    auto snapiter = snaps.begin();
+    while (snapiter != snaps.end()) {
+      snapid_t curr = snapiter->first;
+      if (curr > snapseq_max) {
+        break;
+      }
+      if (curr - last > 1) {
+        // there's a hole between last seen snap and current snap
+        // add a new interval
+        rs.insert(last + 1, curr - last - 1);
+      }
+      last = curr;
+      snapiter++;
+    }
+    if (last != snapseq_max) {
+      rs.insert(last + 1, snapseq_max - last);
+    }
   } else {
     rs = removed_snaps;
   }
 }
 
+// this iterates linearly over snaps map, building deleted snap intervals in rs.
+// it compares that with cached removed snaps (cached) and returns false if any of
+// interval elements in cached doesn't exists in newly removed snaps (rs).
+bool pg_pool_t::build_removed_snaps_and_check_intersect(interval_set<snapid_t>& rs,
```

Comment:
It looks like the code in this function is only valid when `is_pool_snaps_mode()` returns true, so while it may be valid if you're using pool snaps, it doesn't seem valid if you're using self managed snaps.

Reply:
Since `rs` and `cached` are not modified in this method, they can be passed as `const` references.

Reply:
I suspect this whole method is equivalent to the following:
```c++
build_removed_snaps(rs);
return cached.subset_of(rs);
```

Reply:
I've distilled your patch into the following:

https://github.com/ceph/ceph/pull/17820

Reply:
In order for subset_of to be viable, it would need to dynamically choose between sequential search an lower_bound methods, like the intersect_of implementation since https://github.com/ceph/ceph/pull/17088.